services/ns_net/allocation_builder.cc

```diff
@@ -0,0 +1,77 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/services/ns_net/allocation_builder.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+namespace mojo {
+
+AllocationBuilder::AllocationBuilder()
+    : buffer_(nullptr), buffer_length_(0), data_length_(0) {}
+
+AllocationBuilder::~AllocationBuilder() {
+  free(buffer_);
+}
+
+uint32_t AllocationBuilder::Size() const {
+  return data_length_;
+}
+
+static inline uint32_t RoundPowerOfTwo(uint32_t x) {
+  if (x == 0) {
+    return 1;
+  }
+
+  --x;
+
+  x |= x >> 1;
+  x |= x >> 2;
+  x |= x >> 4;
+  x |= x >> 8;
+  x |= x >> 16;
+
+  return x + 1;
+}
+
+bool AllocationBuilder::Append(const uint8_t* data, uint32_t length) {
```

Comment:
Another approach to solving all of these problems is to use the flow control on the data pipe to control how much we read out of the underlying socket.  Then you don't need a buffer, you don't need to worry about oversized allocations, and the flow control is pushed all the way back into the network.

There's no need to introduce another ring buffer.  The data pipe is already a ring buffer.